Added transition handling between the side drawer and chats

kidroca opened the pull request:
@marcaaron 

### Details
Add a loading transition when chat is selected and the side drawer closes

1. Added a fullscreen loading component - just something for the time being - it needs to be revisited with proper design
2. Added `withDrawerState` HOC
3. Disable the comose input component when the drawer is extended - this fixes some issues with the keyboard staying open
  - doesn't work great on Android but has no side effects, the keyboard can still stay open sometimes. It works on iOS and mWeb the keyboard would always close when you open the drawer 

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #2154 fixes #2150

### Tests
1. From the home screen
2. Have a few conversation
3. Select a conversation
4. The conversation should open with a brief loading
  - there should be a fullscreen loading indicator (centered)
  - on physical device sometimes loading happens so fast you might not see the loader. The transition should be smooth though
5. Go back and select another conversation
  - There's shouldn't be any glimpse of the previous conversation 
  - The transition should feel as you are opening a new page

#### The keyboard should be dismissed when you go back to the LHN
- when the keyboard is visible on the screen going back (opening the drawer) should dismiss it
- This should work 100% on iOS and on mWeb (mobile Safari and Chrome) 
- For some reason it doesn't work always work on Android for me
- #2150 is left open due to the Android problem

### QA Steps
Same as the Tests above

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web

https://user-images.githubusercontent.com/12156624/114084920-607b5800-98b9-11eb-8556-124068e56696.mp4

#### Mobile Web

https://user-images.githubusercontent.com/12156624/114086856-d7195500-98bb-11eb-82a0-3223757c8e4c.mov

#### Desktop

https://user-images.githubusercontent.com/12156624/114090730-780a0f00-98c0-11eb-9948-e7325851b1a3.mov

#### iOS (Simulator)

https://user-images.githubusercontent.com/12156624/114086691-98839a80-98bb-11eb-8768-11a96b166c55.mov

#### Android

https://user-images.githubusercontent.com/12156624/114086136-ee0b7780-98ba-11eb-83ee-31f9a08ce10a.mp4

kidroca committed:
refactor: convert ReportScreen to class component

kidroca committed:
refactor: ReportScreen added drawer state

This should be moved to HOC though

kidroca committed:
feat: Loading components

kidroca committed:
feat: ReportScreen add loading animation for content

kidroca committed:
refactor: ReportScreen move loading animation to ReportView

kidroca committed:
refactor: ReportActionsView remove extra calls to initial fetch

kidroca committed:
refactor: ReportScreen extract withDrawerState HOC

kidroca committed:
refactor: withDrawerState implement with useIsDrawerOpen for simplicity

kidroca committed:
refactor: ReportView improve transition using drawer state

kidroca committed:
refactor: ReportActionCompose remove drawer check

This check is no longer needed
1. It didn't work accurately
2. The compose component is unmounted when the drawer is expanded

kidroca commented:
@marcaaron This is my rough sketch of how things can work
I want to share it early on to address any concerns

This is a draft - Please don't pay attention to stuff that needs to be cleaned up. I'm yet to move styles add documentation etc... 

There are a few items I want to ask:
- I've read you don't want to use promises - is the usage here OK - I can replace it with check in componentDidUpdate, but it won't be as elegant as now 
- I need to implement optimistic response - e.g. don't wait for `fetchActions` if the component has some data available to render immediately. I need to think about this but if you have an idea 
- The fullscreen loader might actual not be needed at all. I've tried without it and the transition is still very good. I replaced it with `null` 😄. I can add a sample video next week

So all in all I've still got some work to polish this but what do you say?

kidroca commented on src/components/withDrawerState.js:
```diff
@@ -0,0 +1,25 @@
+
+export default function withDrawerState(WrappedComponent) {
+    const HOC_Wrapper = (props) => {
+        const isDrawerOpen = useIsDrawerOpen();
```
I think I know what you would say about the hook usage, but please check how simple it is compared to this: https://github.com/Expensify/Expensify.cash/pull/2221/commits/8a6df1ba02e2f09fb44517373e6515207eb79b33

Anyway I can swap it for the changes in the diff above

marcaaron replied:
At this point, we are using the hook, but wrapping them in an HOC which makes us wonder _why not_ use the hook. Let's see if we can resolve it [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1617649919372600).

kidroca replied:
We still might need this wrap because class components cannot use hooks directly 
Currently this is only used in `ReportView` which became stateless func component and can use `useIsDrawerOpen` but if I want to add it to `ReportActionCompose` for example, I'll need a HOC

marcaaron replied:
Cool, I think that's fine. And we have updated the style guide.

kidroca commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -175,10 +174,6 @@ class ReportActionCompose extends React.Component {
-        // We want to make sure to disable on small screens because in iOS safari the keyboard up/down buttons will
-        // focus this from the chat switcher.
-        // https://github.com/Expensify/Expensify.cash/issues/1228
-        const inputDisable = this.props.isSmallScreenWidth && Navigation.isDrawerOpen();
```
This check is no longer needed as you would see below the ReportActionCompose is unmounted when the drawer is expanded

kidroca commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -189,9 +188,6 @@ class ReportActionsView extends React.Component {
         Timing.end(CONST.TIMING.SWITCH_REPORT, CONST.TIMING.COLD);
-
-        // Fetch the new set of actions
-        fetchActions(this.props.reportID);
```
Removed fetch calls here. As they are performed in the ReportScreen when reportID changes

There might be other items that can be cleaned up now that the component re-mounts on report change

kidroca commented on src/pages/home/report/ReportView.js:
```diff
@@ -7,37 +7,41 @@ import {addAction} from '../../../libs/actions/Report';
+    /* Is the report view covered by the drawer */
+    isDrawerOpen: PropTypes.bool.isRequired,
+};
 
```
This is no longer needed. And I don't think it brought up any optimizations as all the heavy lifting is performed in the <ReportActionsView />

kidroca committed:
refactor: ReportScreen handle initial state

kidroca committed:
refactor: update copy pasted display name

kidroca committed:
refactor: ReportScreen use reportID from route params

`reportID` comes with a delay when the prop comes from onyx
The user first selects a route in the drawer
The route immediately hides the drawer and reviews the report screen
`reportID` have not yet updated and the old report is visible for a moment

Switching to route params changes the `reportID` the moment the user selects
one of the sidebar chat links

kidroca committed:
refactor: ReportScreen expand the drawer on back navigation

Don't trigger a navigation event as this causes a few unintended re-renders

We don't need to trigger a navigation when can review the drawer from any
chat screen we're currently on

kidroca committed:
feat: ReportActionCompose allow autoFocus to be configurable from parents

kidroca committed:
refactor: ReportView optimizations

Keep the view mounted when the report doesn't change
Unmount the input when the drawer covers the view
Set autoFocus only for bigger screens as example

kidroca committed:
refactor: ReportActionsView remove unnecessary logic around report switching

This is now handled by re-mounting
Remove action sorting performed on each render - apply only when actions change

kidroca committed:
refactor: ReportActionsView remove `needsLayoutCalculation`

This prop no longer exists
The calculation seems to be handled differently and this is leftover

kidroca committed:
refactor: ReportActionCompose remove `withWindowDimensions` - unused

This is no longer needed - functionality extracted outside

kidroca commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -76,29 +76,28 @@ class ReportActionsView extends React.Component {
+        // Todo: this should come from props and depend on isDrawerOutOfTheWay
         // Helper variable that prevents the unread indicator to show up for new messages
         // received while the report is still active
         this.shouldShowUnreadActionIndicator = true;
```
On most chats that I've used you'll still see new messages split line even if the chat is still active, the line will disappear when you interact with the chat

marcaaron replied:
This is getting refactored in another open PR. Which consequently is jumping through a lot of hoops to figure out when the report has been changed (which I believe this PR will simplify).

https://github.com/Expensify/Expensify.cash/pull/1977

kidroca replied:
Oki, what I meant was that at least the initial value could come from outside, because on mobile you can still have an active chat covered by the drawer, so you'd want to see an indicator of where new messages start from when you collapse the drawer

marcaaron replied:
>so you'd want to see an indicator of where new messages start from when you collapse the drawer

The PR I linked changes the general behavior of this marker. I'm not sure there has been any discussion about what to do if the drawer is closed or not. But I also think maybe we don't need to worry about this right now?

kidroca commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -110,12 +109,6 @@ class ReportActionsView extends React.Component {
     }
 
     componentDidUpdate(prevProps) {
-        // We have switched to a new report
-        if (prevProps.reportID !== this.props.reportID) {
-            this.reset(prevProps.reportID);
-            return;
-        }
-
```
No longer needed due to re-mounting

kidroca commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -76,29 +76,28 @@ class ReportActionsView extends React.Component {
     }
 
     shouldComponentUpdate(nextProps, nextState) {
-        if (nextProps.reportID !== this.props.reportID) {
-            return true;
-        }
```
No longer needed due to re-mounting

marcaaron replied:
I'm probably missing something obvious here, but why does this component mount and unmount now whereas before it did not?

kidroca replied:
Answered here: https://github.com/Expensify/Expensify.cash/pull/2221#issuecomment-815943686

kidroca commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -219,9 +209,11 @@ class ReportActionsView extends React.Component {
      */
-    updateSortedReportActions() {
-        this.sortedReportActions = _.chain(this.props.reportActions)
+    updateSortedReportActions(reportActions) {
+        this.sortedReportActions = _.chain(reportActions)
```
Updated usages so this doesn't get called for each render but only when there's new data

kidroca commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -343,7 +333,6 @@ class ReportActionsView extends React.Component {
                     action={item.action}
                     displayAsGroup={this.isConsecutiveActionMadeByPreviousActor(index)}
                     onLayout={onLayout}
-                    needsLayoutCalculation={needsLayoutCalculation}
```
This appears to be no longer used

kidroca commented on src/pages/home/ReportScreen.js:
```diff
@@ -6,49 +6,102 @@ import styles from '../../styles/styles';
+    }
+
+    // Todo: ask why getters aren't on top?
+    get canRenderMainContent() {
+        return this.reportID && !this.state.isLoading && !this.state.error;
+    }
+
+    get reportID() {
+        const params = this.props.route.params;
+        return params.reportID;
+    }
```
Can we use computed getters? 
Usually they will be at the top of the file, but eslint is forcing them here

marcaaron replied:
Hmm, I _think_ we should avoid this for now. While convenient, it's not really very common to see in any of our codebases.

kidroca commented on src/pages/home/ReportScreen.js:
```diff
@@ -6,49 +6,102 @@ import styles from '../../styles/styles';
+            error: null,
+        });
+
+        fetchActions(this.reportID)
+            .catch((error) => {
+                // Todo: what should we do if fetching fails for some reason
+                // Todo: see what we currently do and do the same here
+                this.setState({error});
+                console.error(error);
+            })
+            .finally(() => {
+                console.debug('[ReportScreen] Stop loading: ');
+                this.setState({isLoading: false});
+            });
+    }
```
I'm still fine tuning this.

My end goal is:
If there's already available data to be rendered, we don't need to wait fetching. It should still run and then if there's new data it will auto update

Looking at how currently the app works - there would always be data in Onyx due to `fetchAll` 
So the waiting here can be removed altogether and just usa a brief loading animation to cover the FlatList rendering behind

Flat list rendering need further optimizations - that's the bottleneck on mobile native and why chat actions appear chunk by chunk - struggles with rendering (a lot better on prod build but still...). Nothing that can be done in this ticket though

kidroca committed:
feat: ReportScreen loading view layout transitions

Nicer transition on iOS. I can't see any difference on Android

kidroca committed:
refactor: ReportScreen clean up comments and console.logs

kidroca commented on src/pages/home/ReportScreen.js:
```diff
@@ -1,54 +1,109 @@
+            error: null,
+        });
+
+        fetchActions(this.reportID)
+            .catch((error) => {
+                /* Todo: do something here if the user is going to be left with an empty view
+                *   E.g. Alert like "Oops, we couldn't load this conversation. Please Try again"
+                *     button Retry -> this.fetchReport() */
+                this.setState({error});
+                console.error(error);
+            })
+            .finally(() => {
+                // This adds a small fadeout transition when the loader disappears
+                LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
+                this.setState({isLoading: false});
+            });
```
I'm still fine tuning this.

My end goal is:
If there's already available data to be rendered, we don't need to wait fetching. It should still run and then if there's new data it will auto update

Looking at how currently the app works - there would always be data in Onyx due to fetchAll
So the waiting here can be removed altogether and just usa a brief loading animation to cover the FlatList rendering behind

Flat list rendering need further optimizations - that's the bottleneck on mobile native and why chat actions appear chunk by chunk - struggles with rendering (a lot better on prod build but still...). Nothing that can be done in this ticket though

marcaaron replied:
Makes sense to me. 

>just usa a brief loading animation to cover the FlatList rendering behind

Maybe something like a 300ms delay to start ?

kidroca replied:
Ok, this was the promise usage I was talking about BTW

kidroca commented on src/setup/index.android.js:
```diff
@@ -0,0 +1,8 @@
+import {Platform, UIManager} from 'react-native';
+
+export default function () {
+    // To use layout animations in Android: https://reactnative.dev/docs/layoutanimation
+    if (Platform.OS === 'android' && UIManager.setLayoutAnimationEnabledExperimental) {
+        UIManager.setLayoutAnimationEnabledExperimental(true);
+    }
+}
```
I can drop my commit adding LayoutAnimation. It does look nicer on iOS though

marcaaron replied:
Yeah, I'd say there is plenty to focus on here already. But if we want to propose this as a follow up improvement it's likely people would get on board with the suggestion.

kidroca committed:
refactor: ReportScreen restore reportID as number

There are many prop type warnings when this is string.

marcaaron commented:
>I've read you don't want to use promises - is the usage here OK - I can replace it with check in componentDidUpdate, but it won't be as elegant as now

What specific usage are you asking about here?

>I need to implement optimistic response - e.g. don't wait for fetchActions if the component has some data available to render immediately. I need to think about this but if you have an idea

Not sure I understand this. I think you mean we want to display the component immediately if there is stored data before the network data finishes and updates it again with new data?

>The fullscreen loader might actual not be needed at all. I've tried without it and the transition is still very good. I replaced it with null 😄. I can add a sample video next week

We can always remove this later if it’s not needed so I would say to keep for the time being.

marcaaron commented on src/components/Loading/FullscreenLoading.js:
```diff
@@ -0,0 +1,21 @@
+    </View>
+);
+
+const styles = StyleSheet.create({
```
Just a heads up, we have most styles just in a big object for now and not yet using `StyleSheet.create()` - see `styles.js` file.

marcaaron commented on src/components/Loading/LoadingIndicator.js:
```diff
@@ -0,0 +1,101 @@
+    View,
+    StyleSheet,
+} from 'react-native';
+import Svg, {Circle} from 'react-native-svg';
```
Could we save some time and use the `ActivityIndicator` for now? I appreciate the work done here so far. But we usually try to keep things simple. But if there is a great case for this we will look at whether a custom component can solve some problem and create a new ticket for it.

marcaaron commented on src/pages/home/ReportScreen.js:
```diff
@@ -1,54 +1,109 @@
+        if (!this.reportID) { return; }
+
+        // This adds a small fade in transition when the loader appears
+        LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
```
Probably we can save this for another PR

marcaaron commented on src/pages/home/ReportScreen.js:
```diff
@@ -1,54 +1,109 @@
+                /* Todo: do something here if the user is going to be left with an empty view
+                *   E.g. Alert like "Oops, we couldn't load this conversation. Please Try again"
+                *     button Retry -> this.fetchReport() */
+                this.setState({error});
```
This is pretty much against the initial design that we are going for. That said, we haven't fully thought out how error handling will work. The conversation will happen eventually and maybe our original intentions will be challenged, but for now - if the report doesn't load then it doesn't load.

kidroca replied:
Yep, it would be a rare case to end up in this catch, and even then there might still be some chat history available

kidroca commented on src/pages/home/report/ReportView.js:
```diff
@@ -4,40 +4,53 @@ import PropTypes from 'prop-types';
-                        reportID={this.props.reportID}
-                        key={this.props.reportID}
+                        onSubmit={text => addAction(props.reportID, text)}
+                        reportID={props.reportID}
+                        autoFocus={!props.isSmallScreenWidth}
                     />
                 </SwipeableView>
-                <KeyboardSpacer />
-            </View>
-        );
-    }
+            )}
```
I've just spotted an Android issue coming from this: 
1. Open a chat
2. Focus the input field
3. Press the back arrow to go back to the chat list
You'll get this error: 

I'll debug it tomorrow. Worst case we can leave the field mounted...

kidroca commented on src/pages/home/report/ReportView.js:
```diff
@@ -4,40 +4,53 @@ import PropTypes from 'prop-types';
-                        key={this.props.reportID}
+                        onSubmit={text => addAction(props.reportID, text)}
+                        reportID={props.reportID}
+                        autoFocus={!props.isSmallScreenWidth}
```
I've added this as an example. I can revert it if you want.

kidroca replied:
This will be removed on the next push with updates as confirmed in another thread

marcaaron reviewed (commented):
Just leaving some initial comments. Looks great so far.

tgolen commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -29,6 +27,9 @@ const propTypes = {
     reportID: PropTypes.number.isRequired,
 
+    // Should the input focus on mount
+    autoFocus: PropTypes.bool.isRequired,
```
Just jumping in here for a quick second when I saw this. I think this should be an internal detail of `<TextInputFocusable>`. Is it possible to move it all into there and have it be contained?

kidroca replied:
It already kind of is - `TextInputFocusable` passes props to `TextInput` which is configurable by `autoFocus` 
Do you want to update the prop types definition of `<TextInputFocusable>` to contain `autoFocus` or something else?

This is a flag that is only used when the component mounts and is handled by `TextInput` internally
The default is `false` 

Do you want to add some logic inside `TextInputFocusable` to set a different default value?

Whether a field should `autoFocus` (on mount) should be set from outside as it depends on context/usage. Trying to confine the logic inside `TextInputFocusable` would make it depend on (more and more) external information to make that decision - am I big screen, is something else focused, etc.

Even if `TextInputFocusable` have some opinion on the `autoFocus` prop it should still be overridable from outside

Example: A screen renders 2 (or 10) `TextInputFocusable` components which one would take the focus - it would be straightforward when you can set it from outside

tgolen replied:
OK, I think I misunderstood the usage of this prop slightly, thank you for adding that context. 

My main concern was seeing `autoFocus={!props.isSmallScreenWidth}` and I assumed the prop was only being added to disable autoFocus on small screens. If that was the only application of the prop, then that's why I suggested to move that logic to be contained internally in the component itself.

For the case of multiple components being rendered, I can now understand how it's useful to have it be a prop. 

I have a few suggestions to clear this up a little:

1. Remove the logic you have for the screen width. I see that you added a comment saying it was just an example, and we shouldn't commit code that's just an example.
2. Rename this prop to something which more accurately reflects its usage. Perhaps `shouldFocusOnMount`
3. Add more context to the inline comment for this prop to explain the situation about rendering multiple components and needing to control which one has focus.

kidroca replied:
Allright, I'll add it as a propType on the `TextInputFocusable` as well as it's passed there too due to the prop spread
 
About the name I left it `autoFocus` as it mirrors the react native's own text input behavior and name
- https://reactnative.dev/docs/textinput#autofocus
Also in `React` and html the input attribute name is `autoFocus` (autofocus for html) and it has the same behavior as well
So just for that sake as this behavior (focus on mount) is widely recognized as autoFocus please confirm you want me to change the name

tgolen replied:
OK, `autoFocus` will be all right and the comment does a good enough job explaining what it's for. Thanks!

kidroca committed:
Remove todo note, already addressed elsewhere and could cause confusion

kidroca committed:
refactor: ReportScreen remove error state

kidroca committed:
refactor: ReportScreen replace computed getters with methods

kidroca committed:
Revert "feat: ReportActionCompose allow autoFocus to be configurable from parents"

This reverts commit 2f4a7ea8

kidroca committed:
refactor: Address autoFocus usage and add documentation

kidroca committed:
refactor: extract FullscreenLoading styles

kidroca committed:
Drop: custom LoadingIndicator and use the built in ActivityIndicator

kidroca committed:
drop: Remove layout animations

Additional weight on th PR - These can be proposed as a follow up improvement

kidroca committed:
feat: ReportScreen don't wait for fetch just show a brief loading

kidroca committed:
refactor: ReportActionsView remove `reset` method - no longer needed

Involved logic is already covered on mount and on unmount

kidroca committed:
refactor: ReportScreen move `fetchActions` call back to ReportActionsView

ReportActionsView is already handling all the required fetching so it's better
to just encapsulate that logic in a single place

While the Report Screen can just setup a small transition

kidroca committed:
refactor: ReportScreen replace comment with self explanatory code

kidroca commented:
I don't know why exactly this happens. But pressing on the left side of the chat returns focus on the compose input. I've removed all autofocusing logic whatsoever and it still does

The problem is, we unmount the compose input when the drawer is expanded. So when you click on the Header to go back (which is too on the left of course) you will get a critical error that if dismissed your stuck with unresponsive app. On production this directly force closes the app. Video bellow

**This is Android only**

https://user-images.githubusercontent.com/12156624/113907595-3a34ba00-97de-11eb-87c0-45c24ff8f093.mp4

Video shows: 
- tapping on the left side inside the screen hides and shortly after brings back the keyboard 
- tapping on the right side - closes the keyboard as expected - tapped outside of input
- swiping the drawer open dismissed the keyboard as expected 
- pressing the back button in the header crashes the app because focus tries to return to the input field but the field is immediately unmounted by `isDrawerOutOfTheWay` resolving to `false`. This is unrelated to any of the existing autofocus logic because I disabled it everywhere and the issue would still happen..

I've even tried removing almost every other content with no luck. Things I've tried to figure out what causes this:
- debugging - couldn't trace what is calling/setting focus on the field in js, so it's probably something native
- removed all content from the sidebar/drawer - didn't help
- removed everything inside `ReportView` but just added a plain `{isDrawerOutOfTheWay && <TextInput placeholder="Test..." />}` - no change still tries to focus the input
- created a separate to recreate and post a ticket to react-native or react-native navigation but couldn't recreate the issue in an isolated snack...

I can leave the field mountend and set it as disabled (editable=false), but even then the keyboard stays open (though only on Android) when you go back to the LHN

kidroca committed:
refactor: TextInputFocusable add isDisabled prop to the native component too

kidroca committed:
refactor: ReportActionCompose disable while app focus is away from the input

kidroca commented on src/pages/home/report/ReportView.js:
```diff
@@ -4,40 +4,51 @@ import PropTypes from 'prop-types';
-                        onSubmit={text => addAction(this.props.reportID, text)}
-                        reportID={this.props.reportID}
-                        key={this.props.reportID}
-                    />
-                </SwipeableView>
-                <KeyboardSpacer />
-            </View>
-        );
-    }
+            </SwipeableView>
```
Hiding the Keyboard by unmounting this field turned out to be a big problem on Android: https://github.com/Expensify/Expensify.cash/pull/2221#issuecomment-815093078

"Disabling" the field works on all platforms but Android. The keyboard would still stay open 

Other suggested fixes don't work as well (for Android) 
 - calling `Keyboard.dismiss()` when we go back to LHN, briefly closes the keyboard only to bring it back up when you are at LHN

You'll probably want me to either remove this "fix" altogether or keep it inside ReportActionCompose and don't expose it outside

I've exhausted all my options to deal with this in the current PR. This looks like to be something deeper and needs more investigation. I would ask if it can be handled outside of the current branch in a separate issue. Other than that the work on the transitions is complete

marcaaron replied:
Ok, thanks for all the details. I'll test around a bit while I review. I'm not sure I totally get the problem (but might just need to read your explanation a few more times to get it).

kidroca replied:
If you're testing you can checkout from the previous commit: https://github.com/Expensify/Expensify.cash/pull/2221/commits/7847ffa0b1330f5656ebdb91859eb2cf51412f4e
The code in `ReportView` is setup to unmount the compose input which is how I would get the error when I tap back in the header to go to the LHN

marcaaron replied:
Ah ok I see. Yes, let's just not worry about this for now. As long as everything in this PR is working well I'm happy 🙃

kidroca commented:
Updated

marcaaron commented on src/pages/home/ReportScreen.js:
```diff
@@ -1,54 +1,108 @@
+     * Configures a small loading transition of fixed time and proceeds with rendering available data
+     */
+    prepareTransition() {
+        if (!this.getReportID()) { return; }
```
Style thing. This early return show be on a new line.

kidroca replied:
Ok, it isn't actually needed anymore since we no longer call `fetchActions` here so I'll just remove it

marcaaron commented on src/pages/home/ReportScreen.js:
```diff
@@ -1,54 +1,108 @@
+
+                <View style={[styles.dFlex, styles.flex1]}>
+                    <ReportView
+                        isReady={this.isReadyToDisplayReport()}
```
Would it maybe be simpler to not pass `isReady` here and just have the full screen loader in this component?

marcaaron replied:
I think (but not sure) that we would want to hide the header as well if the report is still loading?

kidroca replied:
Yeah good idea, I had it like that initially but tested different stuff by moving it around.
I noticed the Header information is there right away when you select a conversation and that's why I left it "outside the blanket" but maybe it adds to the overall experience so I'll cover it as well

marcaaron commented on src/pages/home/report/ReportView.js:
```diff
@@ -4,40 +4,51 @@ import PropTypes from 'prop-types';
-                <ReportActionsView
-                    reportID={this.props.reportID}
+function ReportView(props) {
+    const isComposeDisabled = props.isDrawerOpen && props.isSmallScreenWidth;
```
Seems like perhaps this logic should be inside `ReportActionCompose` instead of the `ReportView`? e.g. we are passing `isDrawerOpen` to this component via the HOC and also passing `isDisabled` down. But can this all be contained in `ReportActionCompose` and reduce some logic+ extra props here?

kidroca replied:
Yep, I'm cleaning up stuff and will move this

marcaaron commented on src/pages/home/report/ReportView.js:
```diff
@@ -4,40 +4,51 @@ import PropTypes from 'prop-types';
-            <View style={[styles.chatContent]}>
-                <ReportActionsView
-                    reportID={this.props.reportID}
+function ReportView(props) {
```
For whatever reason our convention is to go with the arrow function style of declaring the stateless function components so can we do this here for consistency?

marcaaron commented on src/pages/home/ReportScreen.js:
```diff
@@ -1,54 +1,108 @@
+            >
+                <HeaderView
+                    reportID={this.getReportID()}
+                    onNavigationMenuButtonClicked={this.props.navigation.openDrawer}
```
More of a question, but is there some reason this is preferable to doing what we had before and using the programmatic navigation?

kidroca replied:
Not really - it's just every `Screen` component receives a handle to the navigation that it belongs to

I prefer to use whatever react native navigation specifically passes to this screen. Also tested if this somehow affects the mWeb routing issue - it doesn't 
I'll revert it so it doesn't get in the way if you plan to do some custom/additional stuff in `Navigation.openDrawer()`

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -83,22 +83,22 @@ class ReportActionsView extends React.Component {
         this.keyboardEvent = Keyboard.addListener('keyboardDidShow', this.scrollToListBottom);
         this.recordMaxAction();
-        fetchActions(this.props.reportID);
+        fetchActions(this.props.reportID).catch(console.error);
```
We should not be catching anything at all here. If we want to catch these errors it should be done in `fetchActions()` but no need to return a promise or handle it here.

kidroca replied:
Sorry It's a habit I have to not leave promises get swallowed. It could sometime hide errors:
![image](https://user-images.githubusercontent.com/12156624/114055443-7547f380-9899-11eb-85dd-fed13bc1c7de.png)

Maybe you could make `fetchActions` return `void` so that it's not used incorrectly

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -365,7 +344,6 @@ class ReportActionsView extends React.Component {
         }
 
         this.setUpUnreadActionIndicator();
```
I think this also can be performed when the component mounts now ? 🤔 

And we should have no need for something like 

`this.shouldShowUnreadActionIndicator`

kidroca replied:
Now that you brought it up I can see. I just focused on the logic around report loading, there are probably other things that can be updated too

marcaaron reviewed (changes requested):
Overall looks great. Have a few suggestions and some questions. Thanks for these changes!

kidroca committed:
refactor: rename FullscreenLoadingIndicator

kidroca committed:
refactor: ReportScreen move the full screen loader to this screen

kidroca committed:
refactor: add `visible` prop to loader

kidroca committed:
refactor: ReportView cleanup unused styles

kidroca committed:
refactor: ReportScreen restore Navigation usage

kidroca commented on src/components/TextInputFocusable/index.native.js:
```diff
@@ -48,6 +57,7 @@ class TextInputFocusable extends React.Component {
                 ref={el => this.textInput = el}
                 maxHeight={116}
                 rejectResponderTermination={false}
+                editable={!this.props.isDisabled}
```
resolving `editable` to false would make the field read only it will also dismiss focus and prevent focusing the field while it's not editable

The "big" issue on Android still happens, but since the view is not unmounted it doesn't crash the app, though the Keyboard will pop out 

It seems the issue doesn't occur all the time though, Today when I started it didn't happen for a while, but then suddenly it started to happen again. It's like tapping in the area on the left that can be swiped to open the drawer somehow triggers this behavior

kidroca commented:
> I'm probably missing something obvious here, but why does this component mount and unmount now whereas before it did not?

https://github.com/Expensify/Expensify.cash/blob/master/src/pages/home/report/ReportView.js#L30-L34

I've moved this `key` prop to the wrapper view. Explicitly setting a `key` would re-mount a view (and it's children) when that key changes. It suggested by React so that you can avoid running error prone checks in didUpdate: 

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

> Keys are usually used for dynamic lists but are also useful here. In our case, we could use the user ID to recreate the email input any time a new user is selected

kidroca committed:
refactor: ReportScreen remove unneeded check

kidroca committed:
refactor: ReportActionsView setup unread indicator during mount

kidroca committed:
refactor: ReportScreen remove navigation prop - unused

kidroca committed:
refactor: ReportActionCompose move logic related to disabling the input here

This will:
- remove focus / hide the keyboard when we go back to the LHN
- prevent focusing on the input while it's covered by the LHN
- enable the field when LHN is collapsed (or permanent)

kidroca committed:
Merge remote-tracking branch 'e.cash/master' into kidroca/report-view-loading-state

kidroca committed:
refactor: ReportScreen restore original navigation handling

kidroca committed:
Merge remote-tracking branch 'e.cash/master' into kidroca/report-view-loading-state

kidroca commented:
Updated 
Added videos and tested on all Platforms
Check the desktop/web transitions - maybe we should skip transition on big screens, as they currently work well.

marcaaron commented:
Thanks, totally missed the addition of the `key`. Makes sense now.

marcaaron reviewed (approved):
Love the code changes here. Thanks!

The full page loader is slightly strange on web/desktop/larger screens, but I think we can address that in a follow up if anyone has major issues with it.

laurenreidexpensify commented:
Assuming no regressions in next week @kidroca this will be paid on 15 April 👍🏽

parasharrajat commented:
Opening the Url expensify.cash directly. shows this screen infinitely.
![image](https://user-images.githubusercontent.com/24370807/114203199-8f570400-9975-11eb-8be6-33416ec19bd4.png)

marcaaron commented:
@kidroca I think this might is because we've switched from using the `currentlyViewedReportID` to the `route.params.reportID`. I missed that change in the PR. So, if we want to do it this way then we'll need to implement the solution discussed here: 

https://expensify.slack.com/archives/C01GTK53T8Q/p1617908395456000

We can either:

1. Implement this fix now and not create a new issue for it
2. Revert this change by going back to using the `currentlyViewedReportID` temporarily and then create a new issue to switch things to route params.

kidroca commented:
The proper fix is to address the navigation it's related to other issues like the URL changing on mWeb, but I can't do it in the scope of the current ticket. If you want me to work on that I'll need a new ticket.

I can revert to using `currentlyViewedReportID` like it did before, though this will degrade the experience on Android. Will post a video

kidroca commented:
Here's how it would look like on Android when I revert to `currentlyViewedReportID` - the prop only updates after a delay. 

https://user-images.githubusercontent.com/12156624/114208662-6418e880-9966-11eb-97bf-95745097de14.mp4

This is an issue with slow AsyncStorage access speed. I don't know how exactly ONYX works, but it might be possible to add a small proxy layer that persists data but keeps in in memory as well, so when it is accessed it can come instantly and don't have to be read from the file system first.
I think the iOS implementation of AsyncStorage probably does something like this as it works instantly there, so this might be a feature that can be implemented for AsyncStorage - configurable in memory proxy for fast access

marcaaron commented:
I leave that up to you @kidroca. It's a regression from what we had before so we must do something for now :) I agree that switching the `reportID` source was probably not in scope so the fix for it doesn't need to be either.

p.s. I think maybe the video shared is to show some other behavior than what you are describing.

kidroca commented:
> p.s. I think maybe the video shared is to show some other behavior than what you are describing.
 
 Yep, sorry 
 
Btw I thought of another patch fix that would better address this at the moment, I'll open a new PR. Should I copy the current PR description there?

marcaaron commented:
>Should I copy the current PR description there?

Ah, no need, linking back to this PR should be enough.

isagoico commented:
We found this issue that looks related to this PR https://github.com/Expensify/Expensify.cash/issues/2327. Not completely sure since the issue is reproducible in production.

roryabraham commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -91,14 +89,13 @@ class ReportActionsView extends React.Component {
-        }
-
         if (!_.isEqual(nextProps.reportActions, this.props.reportActions)) {
+            this.updateSortedReportActions(nextProps.reportActions);
```
More than a year ago that this was added now, but wondering if there was any particular reason this side-effect was put in `shouldComponentUpdate` instead of `componentDidUpdate`?

roryabraham replied:
Is it because you were hoping to avoid having to evaluate `_.isEqual(nextProps.reportAction, this.props.reportActions)` in both `shouldComponentUpdate and `componentDidUpdate`?

marcaaron replied:
Not exactly... `componentDidUpdate()` happens _after_ the `render()` so (as things are now) if we sorted the actions there then we'd also need to trigger an re-render somehow.

Using `shouldComponentUpdate()` with a class property is a way to re-sort actions _before_ the `render()` method is called.

I guess a more current approach could be to recalculate the options in the render with a dependency on `props.reportActions`, but my hook fu is pretty weak tbh 😂 

![DZ-97vzW4AAbcZj](https://user-images.githubusercontent.com/32969087/193165377-09c5801f-1e7b-4976-8c11-6165583f273f.jpeg)

kidroca replied:
> More than a year ago that this was added now, but wondering if there was any particular reason this side-effect was put in `shouldComponentUpdate` instead of `componentDidUpdate`?

Marc already answered, but yes, to capture `sortedActions` before rendering and without triggering a re-render 

> I guess a more current approach could be to recalculate the options in the render with a dependency on `props.reportActions`

If your talking about hooks it would be as simple as 

```js
const sortedActions = useMemo(() => util.sortActions(props.reportActions), [props.reportActions])
```

But then the component would need to be converted from a class to a function based component (to use hooks)

For a class component the only way to sort actions, (only) when the prop changed, is to use `shouldComponentUpdate`

---

Another alternative is to subscribe the view to `reportId` actions, and sort them in the `connection` configuration, so the component always receives sorted actions (they'd still be sorted only if they change)

marcaaron replied:
Yep, that's sort of what I had in mind. But, I think `_.isEqual()` is doing an optimized deep comparison whereas `useMemo()` would only run if the `reportActions` failed a simple object comparison? (so even better)

>Another alternative is to subscribe the view to reportId actions, and sort them in the connection configuration, so the component always receives sorted actions (they'd still be sorted only if they change)

That's pretty interesting. If we are doing a `keyChanged()` then we know the subscriber is getting new data and it's an opportunity to run some logic without having to do the deep compare.

That said, I think we are maybe always creating new objects when merging or setting changes so when calling `keysChanged()` the simple compare in `useMemo()` might be all that we'd need. When keyChanged is called the object reference for something like `props.reportActions` would be pointing to a new object.

marcaaron replied:
This conversation here is slightly related -> https://github.com/Expensify/react-native-onyx/pull/185

roryabraham replied:
> Another alternative is to subscribe the view to reportId actions, and sort them in the connection configuration, so the component always receives sorted actions (they'd still be sorted only if they change)

This is pretty interesting, but I think overall the hook-based approach might be simpler/more accessible to more people in the long term?

Anyways, thanks for the thorough answers here!
